samples/nrf9160/modem_shell/src/ltelc/ltelc.c

```diff
@@ -0,0 +1,505 @@
+		normal_mode_at_cmd =
+			ltelc_sett_normal_mode_at_cmd_str_get(mem_slot_index);
+		len = strlen(normal_mode_at_cmd);
+		if (len) {
+			if (at_cmd_write(normal_mode_at_cmd, response,
+					 sizeof(response), NULL) != 0) {
+				shell_error(
+					shell_global,
+					"Normal mode AT-command from memory slot %d \"%s\" returned: ERROR",
+					mem_slot_index, normal_mode_at_cmd);
+			} else {
+				shell_print(
+					shell_global,
+					"Normal mode AT-command from memory slot %d \"%s\" returned:\n\r %s OK",
+					mem_slot_index, normal_mode_at_cmd,
+					response);
+			}
+		}
+	}
+
+	return 0;
+}
+
+void ltelc_rsrp_subscribe(bool subscribe)
+{
+	ltelc_subscribe_for_rsrp = subscribe;
+	if (shell_global != NULL) {
+		if (ltelc_subscribe_for_rsrp) {
+			/* print current value right away: */
+			shell_print(shell_global, "RSRP subscribed");
+			if (modem_rsrp != LTELC_RSRP_VALUE_NOT_KNOWN) {
+				shell_print(shell_global, "RSRP: %d", modem_rsrp);
+			}
+		} else {
+			shell_print(shell_global, "RSRP unsubscribed");
+		}
+	}
+}
+
+void ltelc_ncellmeas_start(bool start)
+{
+	int ret;
+
+	if (start) {
+		shell_print(
+			shell_global,
+			"Neighbor cell measurements and reporting starting");
+		ret = lte_lc_neighbor_cell_measurement();
+		if (shell_global != NULL) {
+			if (ret) {
+				shell_error(
+					shell_global,
+					"lte_lc_neighbor_cell_measurement() returned err %d",
+					ret);
+				shell_error(
+					shell_global,
+					"Cannot start neigbor measurements");
+			}
+		}
+	} else {
+		ret = lte_lc_neighbor_cell_measurement_cancel();
+		if (shell_global != NULL) {
+			if (ret) {
+				shell_error(
+					shell_global,
+					"lte_lc_neighbor_cell_measurement_cancel() returned err %d",
+					ret);
+			} else {
+				shell_print(
+					shell_global,
+					"Neighbor cell measurements and reporting stopped");
+			}
+		}
+	}
+}
+
+#define AT_MDM_SLEEP_NOTIF_START "AT%%XMODEMSLEEP=1,%d,%d"
+#define AT_MDM_SLEEP_NOTIF_STOP "AT%XMODEMSLEEP=0"
+void ltelc_modem_sleep_notifications_subscribe(uint32_t warn_time_ms,
+					       uint32_t threshold_ms)
+{
+	char buf_sub[48];
+	int err;
+
+	snprintk(buf_sub, sizeof(buf_sub), AT_MDM_SLEEP_NOTIF_START,
+		 warn_time_ms, threshold_ms);
+
+	err = at_cmd_write(buf_sub, NULL, 0, NULL);
+	if (err) {
+		shell_error(
+			shell_global,
+			"Cannot subscribe to modem sleep notifications, err %d",
+			err);
+	} else {
+		shell_print(shell_global,
+			    "Subscribed to modem sleep notifications");
+	}
+}
+
+void ltelc_modem_sleep_notifications_unsubscribe(void)
+{
+	int err;
+
+	err = at_cmd_write(AT_MDM_SLEEP_NOTIF_STOP, NULL, 0, NULL);
+	if (err) {
+		shell_error(shell_global,
+			    "Cannot stop modem sleep notifications, err %d",
+			    err);
+	} else {
+		shell_print(shell_global,
+			    "Unsubscribed from modem sleep notifications");
+	}
+}
+
+#define AT_TAU_NOTIF_START "AT%%XT3412=1,%d,%d"
+#define AT_TAU_NOTIF_STOP "AT%%T3412=0"
+void ltelc_modem_tau_notifications_subscribe(uint32_t warn_time_ms,
```

Comment:
mosh shell command "ltelc tau --subscribe" with optional params for warn_time and threshold was planned to be a command that can be used in testing and in a run time from mosh to subscribe for TAU notifications, instead of compile time: CONFIG_LTE_LC_TAU_PRE_WARNING_NOTIFICATIONS.

This needs additional testing with custom params but when looking at current NCS LTE_LC code it seems that this isn't working with custom parameters, at least warn_time as a custom, i.e. other than CONFIG_LTE_LC_TAU_PRE_WARNING_TIME_MS.

Custom warn_time_ms seems to be causing expected LTE_LC_EVT_TAU_PRE_WARNING not to come. There seems to be that condition in lib lte_lc.c:
https://github.com/nrfconnect/sdk-nrf/blob/master/lib/lte_link_control/lte_lc.c#L321
@jtguggedal, what do you think, would there be a possibility to extend the functionality of LTE_LC TAU events to these needs for runtime custom params?

Reply:
I think it was done this way to only propagate pre-warnings and not the unsolicited responses when starting T3412 and when it is zero. If the check is removed, those will also be propagated as if they were pre-warnings. The latter can be solved by ensuring we don't propagate if time parameter is zero. There's usually a good chance that T3412 is much larger than the typical warning time, so we could make some assumptions on what other times to forward, and in most cases we would probably be right.

A more robust solution would be for lte_lc to have an API for setting the warning time and threshold. Then we could store the warning time value in a variable and compare to that. But we'd still have issues when TAU notifications are set directly with AT commands.
What do you think @simensrostad ?

Reply:
Having a separate API to do this sounds like the best option. But, as Jan Tore mentioned, this pairs poorly with TAU notifications that are set directly via `AT_CMD`, potentially breaking applications that depend on TAU pre-warnings to schedule data transmission. I'm not sure what benefits a real application could draw from setting this run-time though. IMO we should implement this properly and to do this we would need more information in the `AT%XT3412` response to correctly filter out the associated _events_.

Reply:
OK, for now, I'll remove the option to give custom parameters. Might be good to note that MoSH is actually one of those applications that are giving the XT3412 at-command directly and we are counting still on this event to come. This was done because we don't want to enable TAU notifs only when mosh user says so and not in lte_lc's enable_notifications(), i.e. CONFIG_LTE_LC_TAU_PRE_WARNING_NOTIFICATIONS is NOT enabled.

Reply:
This was discussed further and it was agreed that no need for separate mosh command for TAU if parameters cannot be tuned at runtime, thus I'll remove the mosh tau command and mosh will use plain CONFIG_LTE_LC_TAU_PRE_WARNING_NOTIFICATIONS -feature as it is.